ARCHITECTURE.md

```diff
@@ -0,0 +1,71 @@
+
+## Server
+
+Rojo's server component is divided into a few distinct pieces:
+
+- The web server
+- The CLI
+- The snapshotting system
+
+### The CLI
+
+The Command Line Interface (CLI) of Rojo is the only interface for the program. It's initialized in `main.rs` but is hosted in `src/cli`.
+
+Each command for the CLI is hosted in its own file, with the `mod.rs` file for the `cli` module handling parsing and running each command. The commands are mostly self-contained, though may also interface with Rojo's other code when necessary.
+
+Specifically, they may interface with the web server and snapshotting system.
+
+### The Snapshotting System
+
+To do what it does, Rojo has to do two main things: it must decide how the file system should map to Roblox and then send changes from the file system to the plugin. To accomplish this, Rojo uses what's referred to as snapshots.
+
+Snapshots are essentially a capture of what a given Instance tree looks like at a given time. Once an initial snapshot is computed and sent to the plugin, any changes to the file system can be turned into a snapshot and compared directly against the previous snapshot, which Rojo can then use to make a set of patches that have to be applied by the plugin.
+
+These patches represent changes, additions, and removals to the Roblox tree that Rojo creates and manages.
+
+When generating snapshots, files are 'transformed' into Roblox objects through what's referred to as the `snapshot middleware`. As an example, this middleware takes files named `init.lua` and transforms them into a `ModuleScript` bearing the name of the parent folder. It's also responsible for things like JSON models and `.rbxm`/`.rbxmx` models being turned into snapshottable trees.
+
+Inquiring minds should look at `snapshot/mod.rs` and `snapshot_middleware` for a more detailed explanation.
+
+Because snapshots are designed to be translated into Instances anyway, this system is also used by the `build` command to turn a Rojo project into a complete file. The backend for serializing a snapshot into a file is provided by `rbx-dom`, which is a different project.
+
+### The Web Server
+
+Rojo uses a small web server to forward changes to the plugin. Once a patch is computed by the snapshot system, it's made available via the server's API. Then, the plugin requests regularly, and if a new patch exists, recieves it and applies it in Studio.
+
+The web server itself is very basic, consisting of around half a dozen endpoints. The bulk of the work is performed by either the snapshot system or the plugin, with the web server acting as a middleman.
+
+## The Plugin
+
+This section of the document is left incomplete.
+
+## Data Structures
+
+Rojo has many data structures and their purpose might not be immediately clear at a glance. To alleviate this, they are documented below.
```

Comment:
We should also document the structures in the plugin, like PatchSet. It's particularly important because of how inconsistent things are with camelCase and PascalCase in our structures across the project.

Reply:
We'll need to do this with Plugin and Server headers in this section due to the duplicate naming (ie: ServeSession) making the distinctions critical.

Reply:
Following up
Because this doc is just meant as a high level introduction to Rojo functionality, we don't need to document the details of every plugin structure. I do think PatchSet is still very important to document though, because its the main thing the plugin interacts with during core functionality.

Reply:
We have a dedicated `Plugin` subsection that this can go in if need be, though maybe it's worth giving the Plugin its own document?